### PR TITLE
Dropped Wharf support

### PR DESCRIPTION
**Pull request by quarckster**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -26,8 +26,11 @@ repos:
   rev: v2.34.0
   hooks:
   - id: pyupgrade
-- repo: https://github.com/pre-commit/mirrors-mypy
-  rev: v0.961
+- repo: local
```
Comment (RonnyPfannschmidt) [inline .pre-commit-config.yaml]: please use the additional dependencies for the pre-commit mirros mypy hook

File: pyproject.toml [hunk 2]
```diff
@@ -26,6 +26,10 @@ dependencies = [
 test = [
   "pytest",
   "wait_for",
+  "pre-commit",
```
Comment (RonnyPfannschmidt) [inline pyproject.toml]: those are not test dependencies, i prefer to confine them to pre-commit and/or add a typing extra

File: src/webdriver_kaifuku/__init__.py [hunk 3]
```diff
@@ -39,7 +37,7 @@ def _get_browser_name(webdriver_kwargs: dict) -> str:
     raise ValueError("No browser name specified")
 
 
-@attr.s(auto_attribs=True)
+@define(auto_attribs=True)
```
Comment (RonnyPfannschmidt) [inline src/webdriver_kaifuku/__init__.py:40]: migration of attrs apis should be a separate commit

File: src/webdriver_kaifuku/__init__.py [hunk 4]
```diff
@@ -205,16 +156,16 @@ def is_alive(self) -> bool:
 
     def ensure_open(self) -> WebDriver:
         if self.is_alive:
-            return self.browser
+            return self.browser  # type: ignore
```
Comment (RonnyPfannschmidt) [inline src/webdriver_kaifuku/__init__.py]:
```suggestion
            assert self.browser is not None  # typeguard as is_alive will not communicate as type guard
            return self.browser 
```

File: src/webdriver_kaifuku/__init__.py [hunk 5]
```diff
@@ -205,16 +156,16 @@ def is_alive(self) -> bool:
...
         else:
             return self.start()
 
     def add_cleanup(self, callback: Callable) -> None:
         assert self.browser is not None
         try:
-            cl = self.browser.__cleanup
```
Comment (RonnyPfannschmidt) [inline src/webdriver_kaifuku/__init__.py:215]: wondering if it makes sense to mive those up ito manager to avoid patching driver, also ExitStack is now in the stdlib

**strictly followup**

File: testing/conftest.py [hunk 6]
```diff
@@ -65,13 +65,13 @@ def selenium_container():
 
 
 @pytest.fixture(params=CONFIGS)
-def manager(selenium_container, request: pytest.FixtureRequest):
+def test_data(selenium_container, request: pytest.FixtureRequest):
     from webdriver_kaifuku import BrowserManager, log
 
-    mgr = BrowserManager.from_conf(request.param)
-    mgr.browser_name = request.param["webdriver_options"]["desired_capabilities"][  # type: ignore
+    mgr = BrowserManager.from_conf(request.param)  # type: ignore
```
Comment (RonnyPfannschmidt) [inline testing/conftest.py:71]: i wish pytest had generics there, but it doesnt yet

**Review by RonnyPfannschmidt — CHANGES_REQUESTED** [review]
love the changes, would like to see the distinct topics split into different commits

the followups are nice to have but may be time intensive to slip in, so up to you

File: .pre-commit-config.yaml (hunk 1 above)
Comment (quarckster) [reply .pre-commit-config.yaml]: Is it possible to add the local repo to the pre-commit hook dependencies? local `mypy` and mirrored `mypy` provide different results due to missing `webdriver_kaifuku` in the virtual environment.

Comment (RonnyPfannschmidt) [reply .pre-commit-config.yaml]: then we need to add the source folder in the mypy config (its a pre-commit myoy quirk, we need to set the mypy soruce dirs to include src

Comment (quarckster) [reply .pre-commit-config.yaml]: When I run `pre-commit run mypy --all` I get this result for this config:

```yaml
- repo: https://github.com/pre-commit/mirrors-mypy
  rev: v0.961
  hooks:
    - id: mypy
      language_version: python3
      additional_dependencies:
        - types-attrs
        - types-requests
        - attrs
        - requests
```
```
mypy.....................................................................Passed
- hook id: mypy
- duration: 0.16s

Success: no issues found in 4 source files
```

As you can see pre-commit mypy checked all python files in the repo. But when I run local mypy `mypy . --ignore-missing-imports` I get this result:

```
src/webdriver_kaifuku/__init__.py:160: error: Incompatible return value type (got "Optional[WebDriver]", expected "WebDriver")
Found 1 error in 1 file (checked 4 source files)
```

I believe we don't need to add the source folder but `webdriver_kaifuku` package into additional dependencies:

```yaml
- repo: https://github.com/pre-commit/mirrors-mypy
  rev: v0.961
  hooks:
    - id: mypy
      language_version: python3
      additional_dependencies:
        - types-attrs
        - types-requests
        - attrs
        - requests
        - <path to webdriver_kaifuku directory>
```

provides the same result as local `mypy`. But it's not portable and not recommended.

Comment (RonnyPfannschmidt) [reply .pre-commit-config.yaml]: https://mypy.readthedocs.io/en/stable/config_file.html#confval-mypy_path sorts exactly this

Comment (quarckster) [reply .pre-commit-config.yaml]: Thanks Ronny! Actually, it's not needed but helped me to find the issue. I needed to include all dependencies from `pyproject.toml`. Sorry for the noise.

File: src/webdriver_kaifuku/__init__.py (hunk 5 above)
Comment (quarckster) [reply src/webdriver_kaifuku/__init__.py:215]: I will do it in the follow-up PR.

**Review by RonnyPfannschmidt — APPROVED** [review]
